src/tuf.rs

```diff
@@ -65,8 +90,118 @@ impl<D: DataInterchange, M: Metadata> SignedMetadata<D, M> {
             _metadata: PhantomData
         })
     }
+
+    pub fn signatures(&self) -> &[Signature] {
+        &self.signatures
+    }
+
+    pub fn signatures_mut(&mut self) -> &mut Vec<Signature> {
+        &mut self.signatures
+    }
+
+    pub fn signed(&self) -> &D::RawData {
+        &self.signed
+    }
+}
+
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct RootMetadata {
+    version: u32,
+    expires: DateTime<Utc>,
+    keys: HashMap<KeyId, PublicKey>,
+    roles: HashMap<Role, RoleKeys>,
+    consistent_snapshot: bool,
+}
+
+impl RootMetadata {
+    pub fn new(
+        version: u32,
+        expires: DateTime<Utc>,
+        keys: HashMap<KeyId, PublicKey>,
+        roles: HashMap<Role, RoleKeys>,
+        consistent_snapshot: bool,
+    ) -> Result<Self> {
+        if version < 1 {
+            let msg = format!("Metadata version must be greater than zero. Found: {}", version);
+            bail!(ErrorKind::IllegalArgument(msg));
+        }
+        Ok(RootMetadata { version, expires, keys, roles, consistent_snapshot })
+    }
+
+    pub fn version(&self) -> u32 {
+        self.version
+    }
+
+    pub fn expires(&self) -> &DateTime<Utc> {
+        &self.expires
+    }
+
+    pub fn keys(&self) -> &HashMap<KeyId, PublicKey> {
+        &self.keys
+    }
+
+    pub fn keys_mut(&mut self) -> &mut HashMap<KeyId, PublicKey> {
+        &mut self.keys
+    }
+
+    pub fn roles(&self) -> &HashMap<Role, RoleKeys> {
+        &self.roles
+    }
+
+    pub fn roles_mut(&mut self) -> &mut HashMap<Role, RoleKeys> {
+        &mut self.roles
+    }
+
+    pub fn consistent_snapshot(&self) -> bool {
+        self.consistent_snapshot
+    }
+
+    pub fn add_key(&mut self, role: Role, keyid: KeyId, pubkey: PublicKey) -> Result<()> {
+        if let Some(mut keys) = self.roles.get_mut(&role) {
+            let _ = keys.keyids.insert(keyid.clone());
+        } else {
+            bail!(ErrorKind::IllegalArgument("role not found".into()));
+        }
+        let _ = self.keys.insert(keyid, pubkey);
+        Ok(())
+    }
+
+    pub fn remove_key(&mut self, role: Role, keyid: &KeyId) -> Result<()> {
```

Comment:
This should also remove the key from the known keys object, if no other roles reference it.